File: csharp/1406.stone-game-iii.cs
/*
 * @lc app=leetcode id=1406 lang=csharp
 *
 * [1406] Stone Game III
 */

using System;

// @lc code=start
public class Solution {
    public string StoneGameIII(int[] stoneValue) {
        int n = stoneValue.Length;
        var dp = new int[n + 1];
        for (int i = n - 1; i >= 0; --i) {
            int first = 0;
            dp[i] = int.MinValue;
            for (int k = 0; k < 3 && i + k < n; ++k) {
                first += stoneValue[i + k];
                dp[i] = Math.Max(dp[i], first - dp[1 + i + k]);
            }
        }
        return dp[0] == 0 ? "Tie" : (dp[0] > 0 ? "Alice" : "Bob");
    }
}
// @lc code=end

File: java/1406.stone-game-iii.java
/*
 * @lc app=leetcode id=1406 lang=java
 *
 * [1406] Stone Game III
 */

// @lc code=start
class Solution {
    public String stoneGameIII(int[] stoneValue) {
        int n = stoneValue.length;
        int[] memo = new int[n];
        for (int i = 0; i < n; ++i) {
            memo[i] = Integer.MIN_VALUE;
        }
        int diff = diff(stoneValue, memo, 0);
        return diff > 0 ? "Alice" : (diff == 0 ? "Tie" : "Bob");
    }

    private static int diff(int[] stoneValue, int[] memo, int i) {
        int n = stoneValue.length;
        if (i == n) {
            return 0;
        }
        int ret = memo[i];
        if (ret != Integer.MIN_VALUE) {
            return ret;
        }
        int acc = 0;
        for (int k = 0; k < 3 && i + k < n; ++k) {
            acc += stoneValue[i + k];
            ret = Math.max(ret, acc - diff(stoneValue, memo, i + k + 1));
        }
        memo[i] = ret;
        return ret;
    }
}
// @lc code=end

